File: Two Pointers/392. Is Subsequence.py
"""392. Is Subsequence
"""


class Solution:
    """Solution class
    """

    def isSubsequence(self, s: str, t: str) -> bool:
        """Function to determine if a string is a subsequence of another 

        Args:
            s (str): subsequence
            t (str): main string

        Returns:
            bool: true if subsequence, false if not
        """

        n = len(s)
        if not s:
            return True

        i: int = 0
        for c in t:
            if s[i] == c:
                i += 1
                if i == n:
                    return True
        
        return False


def main() -> None:
    """Main function
    """

    s: str = "abc"
    t: str = "ahbgdc"

    sol = Solution()
    print(sol.isSubsequence(s, t))


if __name__ == "__main__":
    main()

File: Two Pointers/392. Is Subsequence.cpp
/*Is Subsequence
 */

#include <bits/stdc++.h>

class Solution
{
public:
    bool isSubsequence(std::string s, std::string t)
    {
        int n = s.length();
        if (n == 0) return true;

        int i = 0;
        for (const char &c : t)
        {
            if (s[i] == c)
            {
                i++;
                if (i == n)
                    return true;
            }       
        }
        return false;
    }
};

int main()
{
    std::string s = "abc", t = "ahbgdc";

    Solution sol;
    std::cout << sol.isSubsequence(s, t) << std::endl;

    return 0;
}